addons/addon-base-raas/packages/base-raas-services/lib/data-source/data-source-reachability-service.js

```diff
@@ -0,0 +1,262 @@
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+const _ = require('lodash');
+const Service = require('@aws-ee/base-services-container/lib/service');
+const { allowIfActive, allowIfAdmin } = require('@aws-ee/base-services/lib/authorization/authorization-utils');
+const { processInBatches } = require('@aws-ee/base-services/lib/helpers/utils');
+
+const workflowIds = {
+  bulkCheck: 'wf-bulk-reachability-check',
+  accountStatusChange: 'wf-ds-account-status-change',
+};
+
+class DataSourceReachabilityService extends Service {
+  constructor() {
+    super();
+    this.dependency([
+      'auditWriterService',
+      'dataSourceAccountService',
+      'studyService',
+      'workflowTriggerService',
+      'aws',
+      'roles-only/applicationRoleService',
+    ]);
+  }
+
+  async bulkReach(requestContext, { status }, { forceCheckAll = false } = {}) {
+    status = status || '*';
+    const workflowTriggerService = await this.service('workflowTriggerService');
+    const dsAccountIds = await this._getDsAccountsWithStatus(requestContext, status);
+
+    if (_.isArray(dsAccountIds) && dsAccountIds.length > 0) {
+      await workflowTriggerService.triggerWorkflow(
+        requestContext,
+        { workflowId: workflowIds.bulkCheck },
+        {
+          requestContext,
+          status,
+          dsAccountIds,
+          forceCheckAll,
+        },
+      );
+    }
+    // Write audit event
+    await this.audit(requestContext, {
+      action: 'bulk-check-reachability',
+      body: { status, dsAccountIds },
+    });
+  }
+
+  async reachDsAccount(requestContext, { id, type }, { forceCheck = false } = {}) {
+    const accountService = await this.service('dataSourceAccountService');
+    const dataSourceAccount = await accountService.mustFind(requestContext, { id });
+
+    const prevStatus = dataSourceAccount.status;
+    let newStatus;
+    let statusMsg;
+
+    const { reachable, unreachableAppRoles } = await this._checkDsAccountAvailability(
+      requestContext,
+      dataSourceAccount,
+    );
+
+    if (reachable) {
+      newStatus = 'reachable';
+      statusMsg = '';
+    } else if (prevStatus === 'pending') {
+      newStatus = prevStatus;
+      statusMsg = `WARN|||Data source account ${id} is not reachable yet`;
+    } else {
+      newStatus = 'error';
+      statusMsg = `ERR|||Error getting information from data source account ${id}`;
+    }
+    if (prevStatus !== newStatus || forceCheck) {
+      const workflowTriggerService = await this.service('workflowTriggerService');
+      await workflowTriggerService.triggerWorkflow(
+        requestContext,
+        { workflowId: workflowIds.accountStatusChange },
+        {
+          requestContext,
+          id,
+          type,
+        },
+      );
+    }
+    if (_.isArray(unreachableAppRoles) && unreachableAppRoles.length > 0) {
```

Comment:
Is there a reason why you are checking if the unreachableAppRoles is an array? 
Maybe you just wanted to do !_.isEmpty(unreachableAppRoles)?

Reply:
Yeah, I can do that